testing/idris/APKBUILD

```diff
@@ -0,0 +1,116 @@
+# Contributor: Mitch Tishmack <mitch.tishmack@gmail.com>
+# Contributor: Jakub Jirutka <jakub@jirutka.cz>
+# Maintainer: Mitch Tishmack <mitch.tishmack@gmail.com>
+#
+# Note: This abuilds downloads Haskell dependencies in the prepare phase
+#       without any verification!
+pkgname=idris
+pkgver=1.0
+pkgrel=0
+pkgdesc="A Language with Dependent Types"
+url="http://www.idris-lang.org"
+arch="x86_64 armhf"
+license="BSD3"
+depends="gmp-dev gcc"
+makedepends="ghc cabal libffi-dev ncurses-dev zlib-dev"
+subpackages="$pkgname-doc"
+source="$pkgname-$pkgver.tar.gz::https://github.com/idris-lang/Idris-dev/archive/v$pkgver.tar.gz
+	cabal.config"
+builddir="$srcdir/Idris-dev-$pkgver"
+
+export CABAL_CONFIG="$srcdir/cabal/cabal.config"
+
+cabal_update() {
+	msg "Updating constraints"
+
+	cd "$builddir"
+
+	# Create cabal config and fetch index.
+	HOME="${CABAL_CONFIG%/*}" cabal update
+
+	# Resolve deps and generate fresh cabal.config with version constraints.
+	rm -f cabal.config
+	cabal freeze
+
+	# Add version tag at the first line.
+	sed -i "1i--$pkgver" cabal.config
+
+	mv cabal.config "$startdir"/
+
+	cd "$startdir"
+	abuild checksum
+}
+
+prepare() {
+	default_prepare
+
+	if [ "$(head -n 1 "$srcdir"/cabal.config)" != "--$pkgver" ]; then
+		die "Requirements file is outdated, run 'abuild cabal_update'"
+	fi
+
+	cd "$builddir"
+
+	# Create cabal config and fetch index.
+	HOME="${CABAL_CONFIG%/*}" cabal update
+
+	# Cabal implicitly loads cabal.config from the project's directory.
+	ln -sf "$srcdir"/cabal.config .
+
+	# Create Cabal sandbox that will be used for all subsequent cabal
+	# invocations. This is something like Python virtual env.
+	cabal sandbox init
+
+	# Fetch all dependencies.
+	# XXX: This should always fetch the same versions as defined in
+	# our cabal.config, but Cabal does not verify packages integrity!
```

Comment:
Actually I was wrong, the sha256sum's of the package uploaded to hackage is validated as well as more. The overall process amounts to:
- hackage mirrors have a signing key
- when a new version is uploaded it gets signed, and checked at that point and its sizes added to the 01-index.tar.gz file. Which is a collection of versions+sha256sum's and is also signed
- Cabal knows of the public signing keys used in hackage, so when it downloads anything it compares that with what the signed index.tar.gz file has at download time

Example, if we look at:
https://hackage.haskell.org/01-index.tar.gz

We have for idris/1.0:
```
{"signatures":[],"signed":{"_type":"Targets","expires":null,"targets":{"<repo>/package/idris-1.0.tar.gz":{"hashes":{"md5":"10bd38b0d8bb23ee57d0bd4784f541bd","sha256":"b06852c83e307c9398fa33e031ca1fbe3c4f927496e7f5b00f13b4ed19f82beb"},"length":1595719}},"version":0}}
```

I'm still learning about TUF and its intricacies but as implemented in cabal/hackage there is definitely integrity checking.

https://github.com/haskell/hackage-security/blob/master/README.md